include/core/shape.h

```diff
@@ -133,6 +134,950 @@ class Shape : public AutoDerivativeFunction<dim>
+      {
+        Point<dim> bounding_box_center_one;
+        Point<dim> bounding_box_center_two;
+        double     radius_one = this->bounding_box_half_length.norm();
+        double     radius_two = shape.bounding_box_half_length.norm();
+        // In the case of the plane, the radius of the bounding box is zero, so
+        // we replace it with the radius of the other object.
+        if (radius_one == 0)
+          {
+            radius_one = radius_two;
+          }
+        if (radius_two == 0)
+          {
+            radius_two = radius_one;
+          }
+        if constexpr (dim == 3)
+          {
+            bounding_box_center_one =
+              this->position +
+              this->rotation_matrix * this->bounding_box_center;
+            bounding_box_center_two =
+              shape.position +
+              shape.rotation_matrix * shape.bounding_box_center;
+          }
+        else
+          {
+            bounding_box_center_one =
+              this->position +
+              tensor_nd_to_2d(this->rotation_matrix *
+                              point_nd_to_3d(this->bounding_box_center));
+            bounding_box_center_two =
+              shape.position +
+              tensor_nd_to_2d(shape.rotation_matrix *
+                              point_nd_to_3d(shape.bounding_box_center));
+          }
+        if (((bounding_box_center_one - bounding_box_center_two).norm() -
+             radius_one - radius_two) <= 0)
+          {
+            bounding_box_contact = this->bounding_box_contact(shape);
+          }
+        else
+          {
+            bounding_box_contact = false;
+          }
+      }
+
+    // The following algorithm does a minimization of the level_set obtained by
+    // the intersection of two shapes. A usual gradient descent does not work as
+    // such. This algorithm uses a cartesian search combined with an additional
+    // guess based on the gradient of the level set and a numerical gradient
+    // descent.
+    if (bounding_box_contact)
+      {
+        std::vector<Tensor<1, dim>> search_direction;
+        // Define the cartesian direction search directions.
+        if constexpr (dim == 2)
+          {
+            search_direction.push_back(Tensor<1, dim>({1.0, 0.0}));
+            search_direction.push_back(Tensor<1, dim>({-1.0, 0.0}));
+            search_direction.push_back(Tensor<1, dim>({0.0, 1.0}));
+            search_direction.push_back(Tensor<1, dim>({0.0, -1.0}));
+          }
+        else
+          {
+            search_direction.push_back(Tensor<1, dim>({1.0, 0.0, 0.0}));
+            search_direction.push_back(Tensor<1, dim>({-1.0, 0.0, 0.0}));
+            search_direction.push_back(Tensor<1, dim>({0.0, 1.0, 0.0}));
+            search_direction.push_back(Tensor<1, dim>({0.0, -1.0, 0.0}));
+            search_direction.push_back(Tensor<1, dim>({0.0, 0.0, 1.0}));
+            search_direction.push_back(Tensor<1, dim>({0.0, 0.0, -1.0}));
+          }
```

Comment:
What would be simpler here would be to push back the first four vector and push back the last two only if constexpr dim==3. This would make it more readable since the if would contain only the difference.